lib/curlable.js

```diff
@@ -0,0 +1,137 @@
+
+  this.queriesQueue = null;
+
+  this.nextLineHandler = null;
+
+  const self = this;
+  readStreamByLines(this.spawnedProcess.stdout, (line) => {
+    if (!self.nextLineHandler) {
+      console.error('FATAL: No `nextLineHandler`, the tool' +
+        ' is generating output when it is not expected.');
+      console.error('ERROR: ' + line);
+      process.exit(-1);
+    } else {
+      self.nextLineHandler(line);
+    }
+  });
+};
+
+// Either `queriesQueue` is `null`, and `runQueries()` is not running, or
+// `queriesQueue` is not `null`, `runQueries()` is running, and it's currently
+// processing the head of this queue.
+// This invariant is held across function calls, taking advantage
+// of JavaScript uninterrupted function execution property. -- D.K.
+curlable.prototype.runQueries = function() {
+  const id = generateRandomQueryId();
+  this.logger(id + '\t-\t' + this.queriesQueue.query);
+  const begin = Date.now();
+  const self = this;
+  this.nextLineHandler = (result) => {
+    const end = Date.now();
+    self.nextLineHandler = null;
+    self.logger(id + '\t' + (end - begin) + 'ms\t' + result);
+    self.queriesQueue.callback(result);
+    self.queriesQueue = self.queriesQueue.next;
+    if (self.queriesQueue) {
+      self.runQueries();
+    }
+  };
+  this.spawnedProcess.stdin.write(this.queriesQueue.query + '\n');
+};
+
+curlable.prototype.runQuery = function(query, callback) {
+  if (!this.queriesQueue) {
+    // No open queue. Start one, and kick off `runQueries()`.
+    this.queriesQueue = {
+      query,
+      callback,
+      next: null
+    };
+    this.runQueries();
+  } else {
+    // Query in progress. Chain the call and don't invoke `runQueries()`.
+    var it = this.queriesQueue;
+    while (it.next) {
+      it = it.next;
+    }
+    it.next = {
+      query,
+      callback,
+      next: null
+    };
+  }
+};
+
+curlable.prototype.kill = function() {
+  if (this.spawnedProcess) {
+    kill(this.spawnedProcess.pid);
+  }
+};
+
+// Runs the service, `app` is an `express` instance of the app.
+// Prerequisite: `app.use(bodyParser.text({ type: '*/*' }));` was called.
+curlable.prototype.registerRoutes = function(app, route, port, terminationCallback) {
+  const self = this;
+
+  app.get(route, function(req, res) {
+    const url = port ? ('`http://localhost:' + port + route + '`') : '$URL';
+    res
+      .status(405) // Method not allowed.
+      .send('Need POST (`curl -d ... ' + url + '`) or' +
+        ' DELETE (`curl -X DELETE ...`).\n');
+  });
+
+  app.post(route, function(req, res) {
+    self.runQuery(req.body, (result) => {
+      res.header('Access-Control-Allow-Origin', '*').send(result + '\n');
+    });
+  });
+
+  app.delete(route, function(req, res) {
+    res.send('Terminating.\n');
+    self.logger('DELETE request received. Stopping the external service.');
+    self.kill();
+    self.logger('External service stopped.');
+    if (terminationCallback) {
+      terminationCallback();
+    }
+  });
+};
```

Comment:
Overall, the fewer public-facing APIs (this class methods), the better. I'd expect this class to have a constructor (taking all the options to spin up the service) and a destructor (a destroy method which cleans up everything), and maybe a getState method which would tell if the service is up or the DELETE was called to shut it off.

Reply:
I'd prefer to either leave it like this, or receive more precise instructions re. what's The Node Way.